ENH add an inset_axes_from_bounds to the axes class

jklymak opened the pull request:
## PR Summary

### EDIT:  

Name bikeshedding:

Proposal after comments below:

```
Axes.inset_axes_from_bounds
Axes.indicate_inset_bounds
Axes.indicate_inset_zoom
```

`rect` was too ambiguous, `lbwh` too obscure, `bounds` is already used for bboxes (as is extents for lbrt)...

Comments on names still welcome...

### Old PR descr 



Adds `ax.inset_axes_from_rect`.  Pretty straight forward, but note that the `transData` one moves under zoom and pan.  

Note that I chose not to go with the `bbox_to_anchor`, and `loc` approach.  I understand that stuff, but it seems needlessly complicated.  Maybe as a refinement.  

```python 
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
ax.plot(range(10))
axin1 = ax.inset_axes_from_rect([0.8, 0.1, 0.15, 0.15])
axin2 = ax.inset_axes_from_rect([5, 7, 2.3, 2.3], transform=ax.transData)
plt.show()

```
![figure_1](https://user-images.githubusercontent.com/1562854/38630101-c7d5def6-3d6a-11e8-9c76-4e4fac87bbbe.png)


Getting closer:  This is a duplication of https://matplotlib.org/gallery/axes_grid1/inset_locator_demo2.html#sphx-glr-gallery-axes-grid1-inset-locator-demo2-py


```python
import matplotlib.pyplot as plt

import numpy as np

def get_demo_image():
    from matplotlib.cbook import get_sample_data
    import numpy as np
    f = get_sample_data("axes_grid/bivariate_normal.npy", asfileobj=False)
    z = np.load(f)
    # z is a numpy array of 15x15
    return z, (-3, 4, -4, 3)

fig, ax = plt.subplots(figsize=[5, 4], constrained_layout=True)
Z, extent = get_demo_image()
Z2 = np.zeros([150, 150], dtype="d")
ny, nx = Z.shape
Z2[30:30 + ny, 30:30 + nx] = Z

# extent = [-3, 4, -4, 3]
ax.imshow(Z2, extent=extent, interpolation="nearest",
          origin="lower")

axins = ax.inset_axes_from_rect([0.5, 0.5, 0.47, 0.47])
axins.imshow(Z2, extent=extent, interpolation="nearest",
          origin="lower")
# sub region of the original image
x1, x2, y1, y2 = -1.5, -0.9, -2.5, -1.9
axins.set_xlim(x1, x2)
axins.set_ylim(y1, y2)
axins.set_xticklabels('')
axins.set_yticklabels('')

ax.zoom_inset_indicator(axins)
fig.canvas.draw()
plt.show()
```

![figure_1](https://user-images.githubusercontent.com/1562854/38790106-ab80420a-40f3-11e8-81ba-e2fb60825460.png)

EDIT: methods renamed `inset_axes_rect` to allow other APIs in the future.


## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

ImportanceOfBeingErnest commented:
In how far is it bearable or confusing for users to have 

* an `axes_grid1.inset_locator.inset_axes` which creates a `axes_grid1.parasite_axes.AxesHostAxes`
* an `axes.Axes.inset_axes` which creates a `.axes.Axes` ?

(I could imagine the problem arising because people are used to have the same function at different places, e.g. `plt.plot` vs. `axes.Axes.plot`)

jklymak commented:
`mplot3d` call to `plot` objects return a different object than core `plot`.  I don't personally think it should be a problem...

jklymak commented:
I am done w/ this for now, and it can be reviewed if anyone has any suggestions.  (note looks like a spare debug change got into the PR, but I'll remove after/during review).

anntzer commented on examples/mplot3d/lines3d.py:
```diff
@@ -23,7 +23,7 @@
 y = r * np.cos(theta)
 
-ax.plot(x, y, z, label='parametric curve')
+l = ax.plot(x, y, z, label='parametric curve')
```
remove change

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -78,6 +78,26 @@ def _plot_args_replacer(args, data):
+    """
+
+    def __init__(self, rect, trans, parent):
+        self._rect = rect
```
just create the transformedbbox here?

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -78,6 +78,26 @@ def _plot_args_replacer(args, data):
+
+    def __call__(self, ax, renderer):
+        rect = self._rect
+        bbox = mtransforms.Bbox.from_bounds(rect[0], rect[1],
```
would typically write `from_bounds(*rect)`, not that it really matters

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +404,207 @@ def legend(self, *args, **kwargs):
+        # This puts the rectangle into figure-relative coordinates.
+        bb = _inset_locator(rect, transform, self)(None, None)
+        # if not set, set to the same zorder as legend
+        zorder = kwargs.pop('zorder', 5)
```
Legend.zorder, perhaps?

jklymak replied:
Well, I don't think I want it *tied* to the legend value.  And legend's value is hard coded i.e. not an rcParam....

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +404,207 @@ def legend(self, *args, **kwargs):
+
+        zorder : number
+            Defaults to 5 (same as `.Axes.legend`).  Adjust higher or lower
+            to change whether it is above or below data plotted on the axes.
```
"... on the parent axes" (or however you call it)

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +404,207 @@ def legend(self, *args, **kwargs):
+        # This puts the rectangle into figure-relative coordinates.
+        bb = _inset_locator(rect, transform, self)(None, None)
+        # if not set, set to the same zorder as legend
+        zorder = kwargs.pop('zorder', 5)
```
kwargs.setdefault("zorder", 5)

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +404,207 @@ def legend(self, *args, **kwargs):
+            coordinates.
+
+        facecolor : Matplotlib color
+            Facecolor of the rectangle (default 'none')
```
full stops at end of lines (same below)

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +404,207 @@ def legend(self, *args, **kwargs):
+            is '0.5'
+
+        alpha : number
+            Transparency of the rectangle and connector lines.  Default 0.5
```
default *is*

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +404,207 @@ def legend(self, *args, **kwargs):
+        -------
+
+        rectangle_patch, connector_lines :
+            `.Patches.Rectagle`, (four-tupple `.Patches.ConnectionPatch`) one
```
Recta*n*gle tu*p*le

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +404,207 @@ def legend(self, *args, **kwargs):
+            # decide which two of the lines to keep visible....
+            pos = inset_ax.get_position()
+            bboxins = pos.transformed(self.figure.transFigure)
+            rectbbox = mtransforms.Bbox.from_bounds(rect[0], rect[1],
```
from_bounds(*rect)

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +404,207 @@ def legend(self, *args, **kwargs):
+        -------
+
+        rectangle_patch, connector_lines :
+            `.Patches.Rectagle`, (four-tupple `.Patches.ConnectionPatch`) one
```
rectangle, tuple

anntzer commented on lib/matplotlib/axes/_base.py:
```diff
@@ -1048,6 +1048,7 @@ def cla(self):
         self.tables = []
         self.artists = []
         self.images = []
+        self.child_axes = []
```
stash them into self.artists?  I'd rather work towards unifying all these containers rather than adding more of them...

jklymak replied:
Seec omment below for why not implimented....

ImportanceOfBeingErnest commented:
@anntzer I think in its current form `axes_grid1.inset_locator.inset_axes` should not simply be pulled over, because it has some design flaws which become apparent once you leave the standard use case.  
You may read #11060 and links therein to see what is confusing about it (and also review if you like ;)). 

So for the core matplotlib, one should probably spend some more thoughts about how to best design such function to be powerful, but not as confusing.

jklymak commented:
@anntzer and @ImportanceOfBeingErnest, thanks for the feedback.  I think the logic for padded boxes is indeed in `.offsets.AnchoredOffsetBox`.  

`.offsets.AnchoredOffsetBox` works from the idea that you want to put the box somewhere semi-automatically inside an axes (or other bbox).  In that case a padding makes some sense.  But I find the whole API hard to deal with if I just want to put a box (or axes) somewhere deterministic on the figure or parent_axes.     So here, I was trying to be more deterministic with a more streamlined API.  

What might make sense is to add a "pad" parameter in appropriate units (probably points, versus pixels), that would offset the new inset axes from the rectangle provided.  Then if you want upper right you do:

```python
w = 0.2, h=0.2
axins = ax.inset_axes([1-w, 1-h, w, h], pad=5)
```

I don't think thats hard to do, though I'll have to check how its done to be robust under figure size changes.  Could have an `hpad` and `wpad` if desired.  I'd tend to be against different pads left and right, up and down...

Not sure how all that would work with set aspect ratios for the child axes.

Would something like that satisfy the need?  

Conversely, I'm not 100% against just moving the `inset_axes` into core.  But I really don't like the API, so felt this would be a decent time to change it. I appreciate some power users have figured out how to make it work, and recent attempts to improve the documentation have been heroic, but...

anntzer commented:
To be clear, I think this should go in without padding support, because it's basically using the same positioning syntax as `text`, which is a good thing.  Padding support should get its own discussion (and I'm not even certain it's something worth the complexity).

jklymak commented:
I guess personally I don't have a problem with padding just being a fraction of the axes size (i.e. I can hard code it into `rect`), but I can see why it is nice to have in physical units for consistency between different sized-axes.  But I'm more than happy to leave that for a future PR, either as a new method or by enhancing the API of these methods.  I don't think there is anything architecturally that needs to change to make that work.

ImportanceOfBeingErnest commented:
I'm thinking more in terms of ease of use. One can always argue that if people want to have full control they can reach their goal via the axes_grid1 inset or custom transforms etc.
So I'm comparing to `legend` here, because that is similar in the sense that you have a box with some extra information in. You may call `legend()` without any argument and get a nice legend. This has some padding to the outside and if you don't like the position you'd add `loc=4` at most. Apart it's still deterministic and fully customizable, using all the remaining arguments. 

In that sense, something like
```
def inset_axes(loc=3, pad=0.5, width=0.4, height=0.3, box=None, transform=None):
```
which can be called just as  `ax.inset_axes()` could be more desireable.

anntzer commented:
I think "loc=number" is a **terrible** API.  **At least** that should be using strings (which are also allowed with legend, but I would want new APIs to only use strings).

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -78,6 +78,24 @@ def _plot_args_replacer(args, data):
                          "multiple plotting calls instead.")
 
 
+class _inset_locator(object):
```
would just make this a function
```
def _make_inset_locator(rect, trans, parent):
    rect = ...

    def inset_locator(ax, renderer):
        ...
        return bb

    return inset_locator
```
Avoids having to manipulate state yourself.

Not a big deal though.

jklymak replied:
I didn't know the scope would be OK if you did that...

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +402,208 @@ def legend(self, *args, **kwargs):
+        """
+        if transform is None:
+            transform = self.transAxes
+        label = kwargs.pop('label', 'inset_axes')
```
would tend to make this and zorder keyword-only args (see my kwonly-args PR for example).  again, not a big deal.

jklymak replied:
Ah, I see why I ignored you here - this is pretty clear kwarg passing down to axes.  I don't want to pull individual kwargs out for documentation reasons that are just passed down, but on the other hand I do want to set the default to something different than what Axes would set it to by itself.  Moreover, 99% of the time, no one wants to define the label...

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +402,208 @@ def legend(self, *args, **kwargs):
+        label = kwargs.pop('label', 'inset_axes')
+
+        # This puts the rectangle into figure-relative coordinates.
+        bb = _inset_locator(rect, transform, self)(None, None)
```
just create axes_locator below first and reuse it here.

jklymak commented:
@ImportanceOfBeingErnest to carry on the API discussion (somewhat orthogonal to this PR).

While it *does* appear a few times in the code base, I think that an API where one kwarg trumps another is in-elegant and inherently confusing as you don't know easily which kwarg trumps which.  Yes, good docs help w/ this, but its suboptimal.  To my mind, if you want to have the API change when using a different kwarg that then ignores other kwargs, maybe you are better off to just have a new method.  

In this case, I could easily see adding a `ax.inset_axes_loc(loc, width, height)` that somehow used loc to place an axes, but I'd sure like a less ambiguous, more consistent, API than the existing one.

jklymak commented on lib/matplotlib/axes/_base.py:
```diff
@@ -1808,6 +1808,20 @@ def add_artist(self, a):
         self.stale = True
         return a
 
+    def add_child_axes(self, ax):
```
That doesn't work for some reason.    I've not entirely tracked down, but if I just make it a bare artist I get:

```
    @axes.setter
    def axes(self, new_axes):
        if (new_axes is not None and self._axes is not None
                and new_axes != self._axes):
>           raise ValueError("Can not reset the axes.  You are probably "
                             "trying to re-use an artist in more than one "
                             "Axes which is not supported")
E           ValueError: Can not reset the axes.  You are probably trying to re-use an artist in more than one Axes which is not supported
```

anntzer replied:
Ah, indeed... fair enough.

anntzer replied:
Now that you bypassed the setter to set `.axes`, perhaps this can be revisited?

anntzer replied:
Dunno, this passes the tests for me?
```
diff --git a/lib/matplotlib/axes/_base.py b/lib/matplotlib/axes/_base.py
index b66f1dbe0..35b3db9e0 100644
--- a/lib/matplotlib/axes/_base.py
+++ b/lib/matplotlib/axes/_base.py
@@ -1048,7 +1048,6 @@ class _AxesBase(martist.Artist):
         self.tables = []
         self.artists = []
         self.images = []
-        self.child_axes = []
         self.mouseover_set = set()
         self._current_image = None  # strictly for pyplot via _sci, _gci
         self.legend_ = None
@@ -1831,15 +1830,13 @@ class _AxesBase(martist.Artist):
 
         This is the lowlevel version.  See `.axes.Axes.inset_axes`
         """
-
         # normally axes have themselves as the axes, but these need to have
         # their parent...
         # Need to bypass the getter...
         ax._axes = self
         ax.stale_callback = martist._stale_axes_callback
-
-        self.child_axes.append(ax)
-        ax._remove_method = self.child_axes.remove
+        self.artists.append(ax)
+        ax._remove_method = self.artists.remove
         self.stale = True
         return ax
 
@@ -4116,7 +4113,6 @@ class _AxesBase(martist.Artist):
         children.append(self._right_title)
         children.extend(self.tables)
         children.extend(self.images)
-        children.extend(self.child_axes)
 
         if self.legend_ is not None:
             children.append(self.legend_)
diff --git a/lib/matplotlib/tests/test_axes.py b/lib/matplotlib/tests/test_axes.py
index b12266a06..9d87e4f11 100644
--- a/lib/matplotlib/tests/test_axes.py
+++ b/lib/matplotlib/tests/test_axes.py
@@ -5734,6 +5734,8 @@ def test_zoom_inset():
 
     # Make the inset_axes...  Position axes co-ordinates...
     axin1 = ax.inset_axes_rect([0.7, 0.7, 0.35, 0.35])
+    assert axin1.axes == ax
+    assert axin1 in ax.artists
     # redraw the data in the inset axes...
     axin1.pcolormesh(x, y, z)
     axin1.set_xlim([1.5, 2.15])
```

jklymak replied:
OK, I see now.  Sorry, I was being dense. 

I'm OK w/ making the lists more generic, I think.  Curious if that is a consensus opinion.

anntzer replied:
I'm strongly in favor of moving towards a single container for all child artists.

efiring replied:
In the call on July 16, I believe @tacaswell expressed the opposite opinion (not in this context), that we should move towards compound Artists with their own child artist lists.

jklymak replied:
...well - I don't know that deciding this needs to be done now.  Whats done here is that the new axes is just added as an artist to the parent.  But its easy to change to a list of child axes.

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -384,6 +409,208 @@ def legend(self, *args, **kwargs):
     def _remove_legend(self, legend):
         self.legend_ = None
 
+    def inset_axes(self, rect, transform=None, zorder=5,
```
should be kw-only (`(self, rect, transform=None, *, zorder=5, ...)`) because passing zorder positionally would just be hard to read (and, from our PoV, make it harder if we want to change the signature).  same below re: inset_rectangle, etc.

jklymak replied:
ok, I'm always confused about positional arguments.... I don't need `transform` to be on either.  So just putting `*` in there makes all the rest of the args kwarg only?  Thats kind ugly ;-)

anntzer replied:
https://www.python.org/dev/peps/pep-3102/

ImportanceOfBeingErnest commented:
Well, there is self-consistency and consistency with the existing API (e.g. that of `legend`). One might sure argue that the legend API is not perfect, but it is stable and people know how to use it. So it somehow would make sense that new methods show some parallelism to that. `text` is a good counter-example here: Why on earth is it `text(x,y,"text")` but `annotate("text", (x,y))`? 

I'm not saying that any of the above *is* or is not  the way to go. E.g. one could also think of positioning insets inside axes in a similar manner as axes inside figures, e.g. through a grid(spec) or some subplot arguments (like 221).  
But if we learned something from all the inset_axes stuff, then it should be that such things need to be well thought through not to end up in a mess. There seems to be no rush in adding *this* `Axes.inset_axes` right *now*, or is there? Maybe it's worth collecting some arguments first as to what such method would be used for and how people like to use it. And also to think about the internals in a more sustainable manner; seen from above, it's not too clear if there should be something like child axes or not and how those should be handled. In short, I'm missing a bit the overall strategy here.

jklymak commented:
@ImportanceOfBeingErnest I'm in no huge rush for this to be merged if we think there are big API issues.  

With respect to adding children axes, the only idea here is that children axes get counted in the bbox calculation for automatic layout.  The other planned use, in my mind, is to add a second x/yscale to an axes instead of using `twinx/y`.  By adding them as children, we make sure they are counted for tight_layout and/or constrained_layout....

jklymak commented:
OK, my arguments re API:

  1. `legend` is not a good parallel because it doesn't have *width* or *height* arguments.  
  2. `bbox_to_anchor` and `bbox_transform` are a hot mess of an API and even worse, an almost intractable spaghetti of code.  They should not be brought into the main library any more than they already have been via `legend` (which is the *only* place it appears in the main API).  
  3. The API proposed here is simple, and in direct parallel with the existing `set/get_position` api of `.Axes`.  

I'm happy to entertain the idea of a `pad` kwarg.  

I'm also happy to entertain the idea of another method `inset_loc_axes()` or some such that accepts a location and *ideally* just calls `inset_axes()` with a thin, easy-to-follow wrapper.  I'm *not* in favour of `bbox_to_anchor` or `bbox_transform`.  If the user wants to place manually, they can just use `inset_axes`!

jklymak commented:
> I think if you allow for pad you actually need something like loc that determines towards which direction to pad. Else, one could only shrink the whole thing by some amount.

I agree, it would have to shrink. 

>  One could also add a parent_ax argument to add_axes, 

I'm not in favour of this as the primary API - the inset axes is a decorator of the parent axes.  I'd expect to see the method to add an inset in the Axes method, not figure...

The `transform` argument is optional, and I expect most users would never use it.

ImportanceOfBeingErnest commented:
Summary of options.

### Inset axes

During the discussion here, several proposals have been brought up as to how an inset_axes API in the core matplotlib could look like. The common ground is to add a method to the `matplotlib.Axes` and let it return another `matplotlib.Axes` object. The following ideas came across: (Note that I used unique names here to be able to distinguish them, those do not need to be the final name(s).)

* **`add_inset_complete`**  
    ```
   def add_inset_complete(loc="upper right", pad=0.5, width=0.4, height=0.3, 
                       box_to_anchor=None, bbox_transform=None):
    ```
    * **Pro**: Most versatile, yet callable without any arguments. Similar API as `legend`. 
    * **Con**: Might be overkill. Not too different from existing `axes_grid1` `inset_axes`. Involves a learning curve (similar to `legend`).

* **`add_inset_rect`** (essentially this PR). Provide a rectangle box and optionally a transform
    ```
    def add_inset_rect(rect=[.6, .5, .4, .3], transform=None):
    ```
    * **Pro**: Ease of use. In parallel to `fig.add_axes` API.
    * **Con**: Inset position depends on axes size. Hard to obtain small offsets to parent axes' frame.

* **`add_inset_rect_with_pad`** Provide a rectangle box and optionally a transform. In addition allow shrinking, such that there is a padding towards the axes, similar to how a legend is placed.
    ```
    def add_inset_rect_with_pad(rect=[.6, .5, .4, .3], transform=None, shrink=0.5):
    ```  
    * **Pro**: Ease of use. Small offsets to parent axes' frame possible.
    * **Con**: Inset position depends on axes size.

* **`add_inset_loc`** Provide a location and width/height. In addition allow shrinking, such that there is a padding towards the axes, similar to how a legend is placed.
    ```
    def add_inset_loc(loc="upper right", width=0.4, height=0.5, pad=0.5):
    ```  
    * **Pro**: Ease of use. Intuitive positioning.
    * **Con**: No custom positions possible.

* **`add_inset_grid`** Provide a location in terms of a grid position, similar to subplotspec. Could equally take a gridspec, which is interpreted in terms of axes.bbox instead of figure.
    ```
    def add_inset_grid(gs=333, **subplotpars):
    ```  
    * **Pro**: Ease of use. Intuitive positioning, similar to subplotspec API
    * **Con**: Custom positions possible, but require a lot of thinking. No positioning outside of axes possible.


<hr>

### Child axes

There is also the question what those functions should do with the created axes. The idea can be to create an additional list attribute of the axes, `Axes.child_axes` which stores any of those created children.

* **Pro**: The insets are easy to obtain and manage; geometry managers can use special rules for child axes.
* **Con**: This adds just another list of artists to the axes which adds to variety of existing lists and makes it hard to see through where which artist is stored and why.

jklymak commented:
Based on the call, `add_inset_rect` was deemed a good start as simple and straightforward, but renamed so that other APIs would be possible in the future.  A pad could be added to `add_inset_rect` and if there is a strong desire, a "loc"-based API could be added as well... (ping @ImportanceOfBeingErnest)

We discussed API consistency a little bit, as this is a continuing issue with the code base.    I think most of the core devs lean towards a simpler API than is evidenced in `legend` and much of the `axes_grid1` toolkit, though I think its acknowledged that this robs us of some slick features of those APIs.

jklymak commented:
I'd quite like to see this merged soon so that I can use the child axes for other things, it at all possible...

jklymak commented:
Rebased....

jklymak commented:
I know I need to suqash this, but any further review needed on this?  @anntzer you had a number of very helpful comment already...?

anntzer commented:
IIRC one thing that somewhat annoyed me was the following (if I'm not mistaken, didn't try recently -- may well be completely wrong):
Normally, one can get the parent Axes (if any) of an Artist with `artist.axes`, i.e. if an Artist appears in some Axes' `.lines`, `.images`, etc. lists, then the artist's `.axes` attribute will be set to that Axes.  But for inset axes, that's not true: their `.axes` attribute points to themselves (as for non-inset axes).  Is there a way to have `.axes` point to the parent Axes (as I think it should)?  Or some other way to get back to the parent Axes from the inset (but I would prefer not adding a custom API for that)?

jklymak commented:
Thanks, I didn't notice that.  I had to directly add to the private list `axin._axes`, but otherwise that seems to work.

timhoffm commented:
I don't have enough background on insets and parts of the existing matplotlib library to really judge the API design.

It's necessary to think this through from the beginning, but I'm afraid that to a certain extend we have to try things out and learn as we go. Therefore, I recommend to start slow with a small set of features, see how that works and gradually expand the API using the experience gathered.

Some random thoughts:

- If you just need this internally for constrained layout (think I read that somewhere but may be wrong), we could start with a private method.
- Alternatively, could we introduce this as "experimental API"? This would get user feedback, but at the same time makes it clear that the API may still change.
- The current implementation of `inset_axes_rect()` seems simple and specific enough to be able to improve the API later on if needed without breaking things badly. This is good.
- Naming: Be sure to carefully choose the names of the methods.
  Generally, it's recommended that method names should contain a verb, e.g. `add_inset_axes_rect` or `create_inset_axes_rect` not just `inset_axes_rect`.
  I know that this is not always the case in the present API. That may be a weak case for not using a verb here (personally, I don't think it is). Anyway, decide conciously.
  `add` itself is used ambigously in the existing API: add an existing object (`add_patch()`) vs. create a new object (`add_subplot`). I would tend to reserve that name for the former case. So I would choose `create_inset_axes_rect` or `make_inset_axes_rect`.

#### Note on pad/shrink
I think it makes sense to handle these things similar to CSS

- pad=5  # all sides
- pad=(2, 5)  # top/bottom, left/right
- pad=(1, 2, 3, 4)  # top, right, bottom, left

You could also add units: `pad=('2px', '5%')`

This is all not required right now (and not part of the PR), but could be a path to add more flexibility without the need of an API change.

Note also that `margin` would be a more fitting term compared to `pad`/`padding`.

jklymak commented on lib/matplotlib/axes/_base.py:
```diff
@@ -1821,6 +1821,25 @@ def add_artist(self, a):
+
+        Returns the added axes.
+
+        This is the lowlevel version.  See `.axes.Axes.inset_axes`
```
Rephrased as 
```
We use this method when we want layout engines to compute the
added axes in the same bbox as the parent.  See `.axes.Axes.inset_axes`
for an example of use.  
```

timhoffm replied:
Sorry, I totally don't get that. :confused:

Is this method only relevant for layout engines? If so, this should be a private method.
If this is generally relevant for a user, the statement on layout engines is rather confusing. Up to here the user just knows that this adds a child axes. To know if I should use this, I either need a description what it does (what does "add" mean) and/or what this is used for.

Stating an example is good.

jklymak replied:
maybe my coffee hadn't kicked in yet...

```python

This method organizes the new axes as a child of the parent, and
        is useful so the child is included in the bounding box
        returned by `.axes.Axes.get_tightbbox` on the parent axes. This is
        particularly useful for automatic layout via ``tight_layout`` or
        ``constrained_layout`` where hierachical organization helps decide
        spacing between axes. See `.axes.Axes.inset_axes_rect` for an example
        of use to make an inset axes for the parent axes.

```

timhoffm replied:
:+1:

jklymak commented:
@timhoffm Thanks a lot for the review.  I think I hit most of your specifics.  In response to your comments: 

- we could add a note about the new fcns being experimental.  
- I see what you are saying about verbs in the method names, but looking at the code base, we don't do that very often.  ie. we do `ax.text`, `ax.arrow`, not `ax.add_text` and `ax.add_arrow`.  I don't wildly object to the verbosity but am not convinced its needed either, unless we think that other verbs acting on the same object are planned.  
- the `add_child_axes` is needed for constrained layout and any other layout engine than needs to know how to calculate the parent bbox.  Moving this from `axes_grid1` has nothing to do with CL.  It is because its a features users often want, and it is a bit strange to point them to a toolbox.  Further the toolbox is somewhat poorly maintained and documented and the code a bit labyrinthian.  The hope here is to make these methods available with less user and maintainer pain.

timhoffm commented:
> we could add a note about the new fcns being experimental.

:+1:  Something along the lines of the message in constrained layout (may be less verbose).

> about verbs ... we don't do that very often.

We don't do this for all the plotting functions. Written explicitly that would be something like `plot_text`, `plot_arrow`. Plotting is sort of the canonical action for an axes, which may justify to leave this out (I doubt that this was a conscious decision, it's probably rather taken from MATLAB). I think adding an inset axes is something more profound and should be a bit more explicit.

That said, I'm in favor of the verb-form, but can also with the noun-only version. My only request is that we think about the API we want (also in the future) and make that decision consciously. Bad naming is a nuisance for the user and changing names in a public API is really difficult.

efiring commented:
@jklymak Naming suggestion: inset_axes_rect -> inset_axes_from_rect

jklymak commented:
Changed name, though not 100% sre the extra verbosity is called for.  Rebased, and I think I added to the API list properly.

jklymak commented:
also think I caught all of @timhoffm doc suggestions....

timhoffm commented on examples/subplots_axes_and_figures/zoom_inset_axes.py:
```diff
@@ -0,0 +1,60 @@
+    f = get_sample_data("axes_grid/bivariate_normal.npy", asfileobj=False)
+    z = np.load(f)
+    # z is a numpy array of 15x15
+    return z, (-3, 4, -4, 3)
```
Just stumbled over this. To make it clear what those magic numbers are use
~~~
extent = (-3, 4, -4, 3)
return z, extent
~~~

jklymak replied:
Thanks, fixed, though I have a sneaky suspicion that I had it that way, and someone suggested I shorten it (ahem, @anntzer :wink:)

anntzer replied:
For once I don't think it's my fault :) (?)

efiring commented:
There is a Travis failure on Python 3.5 having to do with a conflict between the pytest version and the version required for a plugin.  It looks completely unrelated to this PR, and I have no idea how it has arisen.

jklymak commented:
@efiring  Yes, the 3.5 error folks are on it, but haven't found soln yet.  But unrelated to this or other recent PRs...

tacaswell commented on examples/subplots_axes_and_figures/zoom_inset_axes.py:
```diff
@@ -0,0 +1,61 @@
+axins.set_xticklabels('')
+axins.set_yticklabels('')
+
+ax.zoom_inset_indicator(axins)
```
​​highlight_zoom_region
​​outline_zoom_region
​​add_inset_locator
add_zoom_highlight
​​add_zoom_roi (paired with add_inset_roi)

jklymak commented:
### Names as of 18 June:

```
Axes.inset_axes_from_lbwh
Axes.indicate_inset_lbwh
Axes.indicate_inset_zoom
```

ImportanceOfBeingErnest commented:
If the aim of this is to attract users to use this new function, instead of the `axes_grid1.inset_locator.inset_axes` I'm affraid that's not going to happen, simply because `inset_axes_from_lbwh` doesn't sound like a usual public method that makes things easier than what people know as `inset_axes`. And even then, it sounds like you have a choice and could call `inset_axes_from_lbrt` or `inset_axes_from_bbox` as well. 
Indicating the api inside the method's name is quite unusual. It's not like we have `scatter_from_xy` or `legend_from_handlesandlabels` or `set_xticks_from_xticks`. 

It seems the two main points to consider for a name are

1) It should not be `inset_axes`, because that one exists already in a different class
2) It should have a verb inside
3) It should be versatile to be able to add similar methods with other/enhanced functionality later if needed.

Since many nouns are verbs in english, it should be possible to use

* `inset`  - that's a noun and a verb, peoble who need verbs can understand it that way
* `inset_rect`  - that's for distinction between the rectangle and possible other methods, I don't see ambiguity as a problem, most methods that add boxes are ambiguous.
* `frame_inset`  - you may understand *frame* as verb; still one is free to add `localize_inset` (for use with some `loc` parameter) or  `pad_inset` (for use with some padding) or `grid_inset` later
* `inset_frame` (same as above, just reversed, then later use `inset_padframe`, `inset_locframe`)
* `insert_inset` 
* `create_inset` - this has been dropped somehow, but it's still much better than anything with `lbrt` or `lbwh`

jklymak commented:
The `lbwh` was a bigger API discssion during the call.  `rect` etc are ambiguous in name, because no one remembers if its width/height or top/right.  So the long-term idea was to make `lbwh` a standard.  OTOH, I'm not 100% convinced it was more than spitballing at the time :wink:  I'm personally somewhat leery...

ImportanceOfBeingErnest commented:
If you use `rect` people will need to consult the docs in order to know how to specify the rectangle.  
If you use `lbwh` people will need to consult the docs in order to know how the function is named that would create the rectangle.
= no gain whatsoever.

Leaving both out at least helps people to remember the then shorter method name to find it easily in the docs.

jklymak commented:
Quick google search: These axes are definitely called "inset" axes.  So that has to be in the name.  Further, this is an axes, so I think that has to be in the name.  `inset_axes` or `add_inset_axes` are my favourites.  

The reason to add `rect` in the first palce was to leave an API hole for other ways of specifying the inset_axes (ie by location)

jklymak commented:
Thanks @timhoffm

The suggestion for the lbwh stuff came from @efiring and @tacaswell.  Maybe I misunderstood and they only meant the rect argument versus the name of the function.  

However in that conversation I predicted someone would suggest an even funkier new class to specify the rectangle and that was pretty much ruled out as too heavy an API.

ImportanceOfBeingErnest commented:
There are already two classes for specifying rectangles, `matplotlib.patches.Rectangle` and `matplotlib.transform.BBox`. If at all one would sure reuse them instead of creating a new one. 

`matplotlib.transform.BBox` by the way has two static methods, [`from_extents`](https://matplotlib.org/devel/transformations.html#matplotlib.transforms.Bbox.from_extents) and [`from_bounds`](https://matplotlib.org/devel/transformations.html#matplotlib.transforms.Bbox.from_bounds) both taking an 4 element interable as input. One could go for consistency with the naming of those and name this `inset_axes_from_bounds`. Not a big fan of this either, but still better than `lb..` whatever 4 letter combination.

jklymak commented:
@ImportanceOfBeingErnest Thats true - `bounds` is used to mean what `lbwh` means here.  I think the objection to bounds versus extent is that its pretty hard to remember which is which. OTOH they are English names and not obscure acronyms.  

I'll argue that we should use the existing convention, and if someone wants to change the convention in the future, we do the renam then....

Therefore, I've changed the names again:

```python
   Axes.inset_axes_from_bounds
   Axes.indicate_inset_bounds
   Axes.indicate_inset_zoom
```

jklymak commented:
ping @efiring and @timhoffm if the new name suits.  We could discuss on Monday's dev call, except I'll be unavailable.  Happy to chat it out on gitter as well...

jklymak commented:
Rebased and squashed.  Be nice to get this in relatively soon because then I can move onto axes for secondary scales...

jklymak commented:
recycled for CI

jklymak commented:
rebased

jklymak commented:
This just needs one more review to go in as a pretty useful 3.0 feature.  OTOH, understand if it needs to get pushed 6 months...

efiring commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -84,9 +84,37 @@ def _plot_args_replacer(args, data):
+    _parent = parent
+
+    def inset_locator(ax, renderer):
+        bbox = _rect
```
Is this line actually doing anything other than renaming a variable?  I think it is superfluous.

jklymak commented:
Rebased and renamed to plain old `inset_axes` per @efiring.  @timhoffm that may invalidate your approval?  I'll leave it to the maintainers whether this has enough approval...
